core/pkg/store/store.go

```diff
@@ -131,146 +210,180 @@ func (f *Store) String() (string, error) {
+				s.logger.Error(fmt.Sprintf("error deleting flag: %s, %v", key, err))
 				continue
 			}
+
+			s.logger.Debug(
+				fmt.Sprintf(
+					"store resync triggered: flag %s has been deleted from source %s",
+					key, source,
+				),
+			)
+			continue
 		}
 	}
-	f.mx.Unlock()
+	s.mx.Unlock()
 	for key, newFlag := range flags {
-		newFlag.Source = source
-		newFlag.Selector = selector
+		s.logger.Debug(fmt.Sprintf("got metadata %v", metadata))
+
 		newFlag.Key = key
-		storedFlag, _, ok := f.Get(context.Background(), key)
+		newFlag.Source = source
+		newFlag.Priority = priority
+		newFlag.Metadata = mergeMetadata(metadata, newFlag.Metadata)
+
+		// flagSetId defaults to a UUID generated at startup to make our quires isomorphic
+		flagSetId := nilFlagSetId
+		// flagSetId is inherited from the set, but can be overridden by the flag
+		setFlagSetId, ok := newFlag.Metadata["flagSetId"].(string)
 		if ok {
-			if !f.hasPriority(storedFlag.Source, source) {
-				f.logger.Debug(
-					fmt.Sprintf(
-						"not merging: flag %s from source %s does not have priority over %s",
-						key, source, storedFlag.Source,
-					),
-				)
-				continue
-			}
-			if reflect.DeepEqual(storedFlag, newFlag) {
-				continue
-			}
+			flagSetId = setFlagSetId
 		}
-		if !ok {
-			notifications[key] = map[string]interface{}{
-				"type":   string(model.NotificationCreate),
-				"source": source,
-			}
-		} else {
-			notifications[key] = map[string]interface{}{
-				"type":   string(model.NotificationUpdate),
-				"source": source,
+		newFlag.FlagSetId = flagSetId
+
+		raw, err := txn.First(flagsTable, keySourceCompoundIndex, key, source)
+		if err != nil {
+			s.logger.Error(fmt.Sprintf("unable to get flag %s from source %s: %v", key, source, err))
+			continue
+		}
+		oldFlag, ok := raw.(model.Flag)
+		// if we already have a flag with the same key and source, we need to check if it has the same flagSetId
+		if ok {
+			if oldFlag.FlagSetId != newFlag.FlagSetId {
+				_, err = txn.DeleteAll(flagsTable, idIndex, oldFlag.FlagSetId, key)
+				if err != nil {
+					s.logger.Error(fmt.Sprintf("unable to delete flags with key %s and flagSetId %s: %v", key, oldFlag.FlagSetId, err))
+					continue
+				}
 			}
 		}
 		// Store the new version of the flag
-		err := txn.Insert("flags", newFlag)
+		s.logger.Debug(fmt.Sprintf("storing flag: %v", newFlag))
+		err = txn.Insert(flagsTable, newFlag)
 		if err != nil {
-			f.logger.Error(fmt.Sprintf("unable to insert flag %s: %v", key, err))
+			s.logger.Error(fmt.Sprintf("unable to insert flag %s: %v", key, err))
 			continue
 		}
 	}
 
 	txn.Commit()
-	return notifications, resyncRequired
+	return utils.BuildNotifications(oldFlags, flags), resyncRequired
 }
 
-func (f *Store) GetMetadataForSource(source string) model.Metadata {
-	f.mx.RLock()
-	defer f.mx.RUnlock()
-	perSource, ok := f.MetadataPerSource[source]
-	if ok && perSource != nil {
-		return maps.Clone(perSource)
+func mergeMetadata(m1, m2 model.Metadata) model.Metadata {
```

Comment:
I'd call it `patch...(original, patch)` or `overlay(old, new)` as m2 can override some values.